### ISSUE TITLE
sbt 1.0.3 can't overwrite SNAPSHOT artifacts on sonatype snapshot repository

### ISSUE THREAD

**Issue by takezoe** [sbt/sbt#3725]
## steps

Publish configuration:
```scala
publishTo in ThisBuild := {
  if (version.value.trim.endsWith("SNAPSHOT")) Some(Opts.resolver.sonatypeSnapshots)
  else Some(Opts.resolver.sonatypeStaging)
}
```
Plugin configuration:
```scala
addSbtPlugin("com.jsuereth" % "sbt-pgp" % "1.1.0")
```

and run `sbt publishSigned` twice.

## problem

First publishing is succeeded, but second is failed with the following error:

```
[error] java.io.IOException: destination file exists and overwrite == false
[error] 	at org.apache.ivy.plugins.repository.url.URLRepository.put(URLRepository.java:75)
[error] 	at sbt.internal.librarymanagement.ConvertResolver$LocalIfFileRepo.put(ConvertResolver.scala:365)
[error] 	at org.apache.ivy.plugins.repository.AbstractRepository.put(AbstractRepository.java:130)
[error] 	at sbt.internal.librarymanagement.ConvertResolver$ChecksumFriendlyURLResolver.put(ConvertResolver.scala:117)
[error] 	at sbt.internal.librarymanagement.ConvertResolver$ChecksumFriendlyURLResolver.put$(ConvertResolver.scala:104)
[error] 	at sbt.internal.librarymanagement.ConvertResolver$$anonfun$defaultConvert$lzycompute$1$PluginCapableResolver$1.put(ConvertResolver.scala:164)
[error] 	at org.apache.ivy.plugins.resolver.RepositoryResolver.publish(RepositoryResolver.java:216)
[error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$5(IvyActions.scala:492)
[error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$5$adapted(IvyActions.scala:491)
[error] 	at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:789)
[error] 	at scala.collection.Iterator.foreach(Iterator.scala:929)
[error] 	at scala.collection.Iterator.foreach$(Iterator.scala:929)
[error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
[error] 	at scala.collection.IterableLike.foreach(IterableLike.scala:71)
[error] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
[error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
[error] 	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:788)
[error] 	at sbt.internal.librarymanagement.IvyActions$.publish(IvyActions.scala:491)
[error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$3(IvyActions.scala:141)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] 	at sbt.internal.librarymanagement.IvyActions$.withChecksums(IvyActions.scala:154)
[error] 	at sbt.internal.librarymanagement.IvyActions$.withChecksums(IvyActions.scala:148)
[error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$1(IvyActions.scala:141)
[error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$1$adapted(IvyActions.scala:131)
[error] 	at sbt.internal.librarymanagement.IvySbt$Module.$anonfun$withModule$1(Ivy.scala:229)
[error] 	at sbt.internal.librarymanagement.IvySbt.$anonfun$withIvy$1(Ivy.scala:190)
[error] 	at sbt.internal.librarymanagement.IvySbt.sbt$internal$librarymanagement$IvySbt$$action$1(Ivy.scala:70)
[error] 	at sbt.internal.librarymanagement.IvySbt$$anon$3.call(Ivy.scala:77)
[error] 	at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:93)
[error] 	at xsbt.boot.Locks$GlobalLock.xsbt$boot$Locks$GlobalLock$$withChannelRetries$1(Locks.scala:78)
[error] 	at xsbt.boot.Locks$GlobalLock$$anonfun$withFileLock$1.apply(Locks.scala:97)
[error] 	at xsbt.boot.Using$.withResource(Using.scala:10)
[error] 	at xsbt.boot.Using$.apply(Using.scala:9)
[error] 	at xsbt.boot.Locks$GlobalLock.ignoringDeadlockAvoided(Locks.scala:58)
[error] 	at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:48)
[error] 	at xsbt.boot.Locks$.apply0(Locks.scala:31)
[error] 	at xsbt.boot.Locks$.apply(Locks.scala:28)
[error] 	at sbt.internal.librarymanagement.IvySbt.withDefaultLogger(Ivy.scala:77)
[error] 	at sbt.internal.librarymanagement.IvySbt.withIvy(Ivy.scala:185)
[error] 	at sbt.internal.librarymanagement.IvySbt.withIvy(Ivy.scala:182)
[error] 	at sbt.internal.librarymanagement.IvySbt$Module.withModule(Ivy.scala:228)
[error] 	at sbt.internal.librarymanagement.IvyActions$.publish(IvyActions.scala:131)
[error] 	at sbt.Classpaths$.$anonfun$publishTask$4(Defaults.scala:2217)
[error] 	at sbt.Classpaths$.$anonfun$publishTask$4$adapted(Defaults.scala:2215)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:42)
[error] 	at sbt.std.Transform$$anon$4.work(System.scala:64)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:257)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error] 	at sbt.Execute.work(Execute.scala:266)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:257)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:167)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:32)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[error] 	at java.lang.Thread.run(Thread.java:745)
```

## expectation

SNAPSHOT artifacts should be overwritten.

## notes

sbt version: 1.0.3

Some users mentioned about this problem in #3431.

**Comment by pshirshov** [comment]
Not a bug.

It's just a default behaviour which is very easy to change

Full example: https://github.com/pshirshov/izumi-r2/blob/develop/sbt-izumi/src/main/scala/org/bitbucket/pshirshov/izumi/sbt/PublishingPlugin.scala

**Comment by maqdev** [comment]
It is definitely a bug, snapshots should be overwritten by default at sonatype just like they do locally. That's why snapshots exists after all. As for your example I wouldn't say that it's easy, thats looks like a hack.
I should consider it like a workaround and thanks for that 👍

**Comment by mosesn** [comment]
@dwijnand we're running into this too, any idea how we can fix this?  we would like to publish snapshots to maven central, and this seems to be a pretty big impediment.

**Comment by pshirshov** [comment]
Look at my message above, there is a link to a plugin which is fixing the issue.

**Comment by mosesn** [comment]
@pshirshov hmm, OK, I'll try it out.  thanks!

**Comment by channingwalton** [comment]
@mosesn did @pshirshov's plugin work for you? I made those changes to my build but it made no difference for me.

**Comment by Voltir** [comment]
Plugin worked for me with a minor modification to not depend on other plugins in @pshirshov repo. I would rather have this be the default behavior for publishing snapshots though.

**Comment by tindzk** [comment]
Another workaround is to remove the SNAPSHOT packages manually from Sonatype. Click [here](https://oss.sonatype.org/index.html#view-repositories;snapshots~browsestorage), then log in, find the package(s) in the  "Snapshots" pane, right-click and select "Delete".

**Comment by mosesn** [comment]
Yeah, we ended up working around it by curling the delete endpoint in our publish job, but it's not ideal.

@channingwalton: we didn't end up trying @pshirshov's work-around.

**Comment by ffinfo** [comment]
Maybe a little update on this. I found out that in sbt-pgp the overwrite flag is hard coded. This why I did a PR to this plugin:
https://github.com/sbt/sbt-pgp/pull/121

**Comment by maqdev** [comment]
Still happening on `sbt 1.1.1`, easiest workaround I've found is adding these lines to `build.sbt`:

```
// FIXME: remove setting of overwrite flag when the following issue will be fixed: https://github.com/sbt/sbt/issues/3725
publishConfiguration := publishConfiguration.value.withOverwrite(isSnapshot.value)
com.typesafe.sbt.pgp.PgpKeys.publishSignedConfiguration := com.typesafe.sbt.pgp.PgpKeys.publishSignedConfiguration.value.withOverwrite(isSnapshot.value)
publishLocalConfiguration := publishLocalConfiguration.value.withOverwrite(isSnapshot.value)
com.typesafe.sbt.pgp.PgpKeys.publishLocalSignedConfiguration := com.typesafe.sbt.pgp.PgpKeys.publishLocalSignedConfiguration.value.withOverwrite(isSnapshot.value)
```
Credits goes to https://github.com/plokhotnyuk/jsoniter-scala

**Comment by ryan-williams** [comment]
Thanks @maqdev, dropping those lines in my `~/.sbt/1.0/global.sbt` seems like the best work-around until @ffinfo [signs the typelevel CLA](https://github.com/sbt/sbt-pgp/pull/121#issuecomment-365711337) and we get a new sbt-pgp release 😆

**Comment by maqdev** [comment]
This is finally fixed due to `sbt-pgp 1.1.1` is released!
